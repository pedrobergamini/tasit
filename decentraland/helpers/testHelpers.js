import AssetTypes from "@constants/AssetTypes";
const { ESTATE, PARCEL } = AssetTypes;

export const parcel = {
  type: PARCEL,
  id: "0123456789",
  name: "Sample Parcel",
  img: "https://api.decentraland.org/v1/parcels/0/0/map.png",
};

export const estate = {
  type: ESTATE,
  id: "123",
  name: "Sample Estate",
  img: "https://api.decentraland.org/v1/estates/1/map.png",
};

export const estateForSale = {
  id: "1",
  priceMana: 0,
  priceUSD: 0,
  asset: estate,
};

export const parcelForSale = {
  id: "2",
  priceMana: 0,
  priceUSD: 0,
  asset: parcel,
};

export const anAction = {
  getId: async () =>
    "0x1234567890123456789012345678901234567890123456789012345678901234",
};

export default {
  estateForSale,
  parcelForSale,
  parcel,
  estate,
  anAction,
};
