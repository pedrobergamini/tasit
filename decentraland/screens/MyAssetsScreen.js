import React from "react";
import { connect } from "react-redux";
import { setMyAssetsList } from "../redux/actions";
import PropTypes from "prop-types";
import MyAssetsList from "@presentational/MyAssetsList";
import MyAssetsListItem from "@presentational/MyAssetsListItem";
import { listsAreEqual, getContracts } from "@helpers";
import { generateAssetFromId } from "@helpers/decentraland";
import { storeMyAssets } from "@helpers/storage";
import DecentralandUtils from "tasit-sdk/dist/helpers/DecentralandUtils";

export class MyAssetsScreen extends React.Component {
  componentDidMount = async () => {
    const { account, myAssets: assetsFromState, setMyAssetsList } = this.props;
    if (account) {
      const { address } = account;
      const assetsFromBlockchain = await this._getAssetsFromBlockchain(address);
      const shouldUpdate = !listsAreEqual(
        assetsFromBlockchain,
        assetsFromState
      );
      if (shouldUpdate) {
        // TODO: Maybe move these two set/store operations into a single function
        setMyAssetsList(assetsFromBlockchain);
        storeMyAssets(assetsFromBlockchain);
      }
    }
  };

  _getAssetsFromBlockchain = async address => {
    const listOfPromises = await this._getAssetsOf(address);
    const fromBlockchain = await Promise.all([...listOfPromises]);
    // Note: The UI shows land the user bought ordered by purchase date (desc)
    return fromBlockchain.reverse();
  };

  // Note: Returns a list of Promises
  _getAssetsOf = async address => {
    const decentralandUtils = new DecentralandUtils();
    const { getAssetsOf: getLandOf } = decentralandUtils;

    const contracts = getContracts();
    const { estateContract, landContract } = contracts;

    const listOfLand = await getLandOf(address);
    const assets = [];

    for (let land of listOfLand) {
      const { id: assetId, nftAddress, transactionHash } = land;

      let asset = generateAssetFromId(
        estateContract,
        landContract,
        assetId,
        nftAddress
      );

      assets.push({ ...asset, transactionHash });
    }

    return assets;
  };

  _renderItem = ({ item: asset }) => {
    return <MyAssetsListItem asset={asset} />;
  };

  render() {
    const { myAssets } = this.props;

    return (
      <MyAssetsList myAssetsList={myAssets} renderItem={this._renderItem} />
    );
  }
}

MyAssetsScreen.propTypes = {
  myAssets: PropTypes.array.isRequired,
  account: PropTypes.object,
};

const mapStateToProps = state => {
  const { myAssets, accountInfo } = state;
  const { account } = accountInfo;
  const { list: myAssetsList } = myAssets;
  return { myAssets: myAssetsList, account };
};

const mapDispatchToProps = {
  setMyAssetsList,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MyAssetsScreen);
