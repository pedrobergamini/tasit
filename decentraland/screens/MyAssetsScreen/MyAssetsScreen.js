import React from "react";
import { connect } from "react-redux";
import {
  removeFromMyAssetsList,
  appendToMyAssetsList,
  addUserAction,
} from "@redux/actions";
import PropTypes from "prop-types";
import MyAssetsList from "@presentational/MyAssetsList";
import MyAssetsListItem from "@presentational/MyAssetsListItem";
import { listsAreEqual, getContracts, logInfo, logWarn } from "@helpers";
import { generateAssetFromId } from "@helpers/decentraland";
import DecentralandUtils from "tasit-sdk/dist/helpers/DecentralandUtils";
import { SUCCESSFUL } from "@constants/UserActionStatus";

export class MyAssetsScreen extends React.Component {
  componentDidMount = async () => {
    const {
      account,
      removeFromMyAssetsList,
      appendToMyAssetsList,
      addUserAction,
    } = this.props;
    if (account) {
      const { address } = account;

      const boughtAssets = this._getBoughtAssetsFromState();

      const {
        assetsFromBlockchain,
        actionsFromBlockchain,
      } = await this.__getAssetsAndActionsFromBlockchain(address);

      const shouldUpdate = !listsAreEqual(assetsFromBlockchain, boughtAssets);
      if (shouldUpdate) {
        // TODO: Add some UI indication that something unexpected happened
        this._logAssetsInconsistency(assetsFromBlockchain, boughtAssets);

        removeFromMyAssetsList(boughtAssets);
        appendToMyAssetsList(assetsFromBlockchain);
        addUserAction(actionsFromBlockchain);
      }
    }
  };

  _getAssetsAndActionsFromBlockchain = async address => {
    const assetsFromBlockchain = await this._getAssetsFromBlockchain(address);

    let actionsFromBlockchain = {};

    assetsFromBlockchain.forEach(asset => {
      const { actionId, id: assetId } = asset;
      const userAction = { [actionId]: { assetId, status: SUCCESSFUL } };
      actionsFromBlockchain = { ...actionsFromBlockchain, userAction };
    });

    assetsFromBlockchain.forEach(asset => delete asset.actionId);

    return { assetsFromBlockchain, actionsFromBlockchain };
  };

  _getBoughtAssetsFromState = () => {
    const { myAssets: assetsFromState, userActions } = this.props;

    const boughtAssets = assetsFromState.filter(asset => {
      const { id: assetId } = asset;
      const assetAction = userActions.find(
        action => action.assetId === assetId
      );
      return assetAction && assetAction.status === SUCCESSFUL;
    });

    return boughtAssets;
  };

  _logAssetsInconsistency = (fromBlockchain, fromState) => {
    const fromBlockchainIds = fromBlockchain.map(asset => asset.id);
    const fromStateIds = fromState.map(asset => asset.id);

    const addedIds = fromBlockchainIds.filter(id => !fromStateIds.includes(id));
    const removedIds = fromStateIds.filter(
      id => !fromBlockchainIds.includes(id)
    );

    logInfo(
      `Account's assets from blockchain aren't the same as the ones stored on the app.`
    );

    if (addedIds.length > 0)
      logInfo(`Some assets added to MyLandScreen. IDs: [${addedIds}]`);

    if (removedIds.length > 0)
      logWarn(`Some assets removed from MyLandScreen. IDs: [${removedIds}]`);
  };

  _getAssetsFromBlockchain = async address => {
    const listOfPromises = await this._getAssetsOf(address);
    const fromBlockchain = await Promise.all([...listOfPromises]);
    // Note: The UI shows land the user bought ordered by purchase date (desc)
    return fromBlockchain.reverse();
  };

  // Note: Returns a list of Promises
  _getAssetsOf = async address => {
    const decentralandUtils = new DecentralandUtils();
    const { getAssetsOf: getLandOf } = decentralandUtils;

    const contracts = getContracts();
    const { estateContract, landContract } = contracts;

    const listOfLand = await getLandOf(address);
    const assets = [];

    for (let land of listOfLand) {
      const generateAsset = async land => {
        const { id: assetId, nftAddress, transactionHash: actionId } = land;

        const asset = await generateAssetFromId(
          estateContract,
          landContract,
          assetId,
          nftAddress
        );

        return { ...asset, actionId };
      };

      const assetPromise = generateAsset(land);
      assets.push(assetPromise);
    }

    return assets;
  };

  _renderItem = ({ item: assetAndUserAction }) => {
    let { asset, userAction } = assetAndUserAction;
    if (!userAction) userAction = {};
    return <MyAssetsListItem asset={asset} userAction={userAction} />;
  };

  render() {
    const { myAssets, userActions } = this.props;

    return (
      <MyAssetsList
        myAssets={myAssets}
        userActions={userActions}
        renderItem={this._renderItem}
      />
    );
  }
}

MyAssetsScreen.propTypes = {
  myAssets: PropTypes.array.isRequired,
  account: PropTypes.object,
  userActions: PropTypes.object.isRequired,
  removeFromMyAssetsList: PropTypes.func.isRequired,
  appendToMyAssetsList: PropTypes.func.isRequired,
  addUserAction: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  const { myAssets, accountInfo, userActions } = state;
  const { account } = accountInfo;
  const { list: myAssetsList } = myAssets;
  return { myAssets: myAssetsList, account, userActions };
};

const mapDispatchToProps = {
  removeFromMyAssetsList,
  appendToMyAssetsList,
  addUserAction,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MyAssetsScreen);
