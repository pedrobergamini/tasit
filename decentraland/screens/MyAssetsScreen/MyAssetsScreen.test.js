import React from "react";
import { shallow } from "enzyme";
import { MyAssetsScreen } from "./MyAssetsScreen";
import { parcel, parcelUserAction } from "@helpers/testHelpers";

describe("ListLandForSaleScreen", () => {
  describe("renders the component", () => {
    const removeFromMyAssetsList = () => {};
    const appendToMyAssetsList = () => {};
    const addUserAction = () => {};

    it("without assets", async () => {
      const myAssets = [];
      const userActions = {};
      expect(
        shallow(
          <MyAssetsScreen
            myAssets={myAssets}
            userActions={userActions}
            removeFromMyAssetsList={removeFromMyAssetsList}
            appendToMyAssetsList={appendToMyAssetsList}
            addUserAction={addUserAction}
          />
        )
      ).toMatchSnapshot();
    });
    it("with assets", async () => {
      const myAssets = [parcel];
      const userActions = { ...parcelUserAction };
      expect(
        shallow(
          <MyAssetsScreen
            myAssets={myAssets}
            userActions={userActions}
            removeFromMyAssetsList={removeFromMyAssetsList}
            appendToMyAssetsList={appendToMyAssetsList}
            addUserAction={addUserAction}
          />
        )
      ).toMatchSnapshot();
    });
  });
});
