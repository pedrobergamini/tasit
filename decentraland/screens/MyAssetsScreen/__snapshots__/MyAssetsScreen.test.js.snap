// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListLandForSaleScreen renders the component with assets 1`] = `
<MyAssetsList
  myAssets={
    Array [
      Object {
        "id": "0123456789",
        "img": "https://api.decentraland.org/v1/parcels/0/0/map.png",
        "name": "Sample Parcel",
        "type": "PARCEL",
      },
    ]
  }
  renderItem={[Function]}
  userActions={
    Object {
      "0x1234567890123456789012345678901234567890": Object {
        "assetId": "0123456789",
        "status": "SUCCESSFUL",
      },
    }
  }
/>
`;

exports[`ListLandForSaleScreen renders the component without assets 1`] = `
<MyAssetsList
  myAssets={Array []}
  renderItem={[Function]}
  userActions={Object {}}
/>
`;
